internal/core/src/query/CachedSearchIterator.cpp

```diff
@@ -0,0 +1,355 @@
+    if (iterators_.empty()) {
+        return;
+    }
+
+    if (iterators_.size() != nq_ * num_chunks_) {
+        PanicInfo(ErrorCode::UnexpectedError,
+                  "Iterator size mismatch, expect %d, but got %d",
+                  nq_ * num_chunks_,
+                  iterators_.size());
+    }
+
+    ValidateSearchInfo(search_info);
+
+    search_result.total_nq_ = nq_;
+    search_result.unity_topK_ = batch_size_;
+    search_result.seg_offsets_.resize(nq_ * batch_size_);
+    search_result.distances_.resize(nq_ * batch_size_);
+
+    for (size_t query_idx = 0; query_idx < nq_; ++query_idx) {
+        auto rst = GetBatchedNextResults(query_idx, search_info);
+        WriteSingleQuerySearchResult(
+            search_result, query_idx, rst, search_info.round_decimal_);
+    }
+}
+
+void
+CachedSearchIterator::ValidateSearchInfo(const SearchInfo& search_info) {
+    if (!search_info.iterator_v2_info_.has_value()) {
+        PanicInfo(ErrorCode::UnexpectedError,
+                  "Iterator v2 SearchInfo is not set");
+    }
+
+    auto iterator_v2_info = search_info.iterator_v2_info_.value();
+    if (iterator_v2_info.batch_size != batch_size_) {
+        PanicInfo(ErrorCode::UnexpectedError,
+                  "Batch size mismatch, expect %d, but got %d",
+                  batch_size_,
+                  iterator_v2_info.batch_size);
+    }
+}
+
+std::optional<CachedSearchIterator::DisIdPair>
+CachedSearchIterator::GetNextValidResult(
+    const size_t iterator_idx,
+    const std::optional<float>& last_bound,
+    const std::optional<float>& radius,
+    const std::optional<float>& range_filter) {
+    auto& iterator = iterators_[iterator_idx];
+    while (iterator->HasNext()) {
+        auto result = ConvertIteratorResult(iterator->Next());
+        if (IsValid(result, last_bound, radius, range_filter)) {
+            return result;
+        }
+    }
+    return std::nullopt;
+}
+
+// TODO: Optimize this method
+void
+CachedSearchIterator::MergeChunksResults(
+    size_t query_idx,
+    const std::optional<float>& last_bound,
+    const std::optional<float>& radius,
+    const std::optional<float>& range_filter,
+    std::vector<DisIdPair>& rst) {
+    auto& heap = chunked_heaps_[query_idx];
+
+    if (heap.empty()) {
+        for (size_t chunk_id = 0; chunk_id < num_chunks_; ++chunk_id) {
+            const size_t iterator_idx = query_idx + chunk_id * nq_;
+            if (auto next_result = GetNextValidResult(
+                    iterator_idx, last_bound, radius, range_filter);
+                next_result.has_value()) {
+                heap.emplace(iterator_idx, next_result.value());
+            }
+        }
+    }
+
+    while (!heap.empty() && rst.size() < batch_size_) {
+        const auto [iterator_idx, cur_rst] = heap.top();
+        heap.pop();
+
+        // last_bound may change between NextBatch calls, discard any invalid results
+        if (!IsValid(cur_rst, last_bound, radius, range_filter)) {
```

Comment:
so v2 iterator will not return better results compared to former iterations page?

Reply:
Not in this stage, the next one will try to take care of this.